### ISSUE TITLE
Allow extension type to implement its representation type.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#3397]
TL;DR: I don't think I can explain to users why the pattern `extension type E(T _) implements T { moreMembers() }` is not allowed if `T` is an extension type, because there is no reason to prevent it. It's a valid and sound pattern, which today is only allowed if `T` is not an extension type.

## Proposal

Change the rule for when an extension type declaration can implement another extension type to:

> Assume that an extension type declaration _DV_ named `Name` has
> representation type `R`, and that the extension type `V1` with
> declaration _DV1_ is a superinterface of _DV_ (*note that `V1` may
> have some actual type arguments*).  Assume that `S` is the
> instantiated representation type corresponding to `V1`. A compile-time
> error occurs if `R` is not a subtype of `S` and `R` is not a subtype of _DV1_.

The only change is the addition of "and `R` is not a subtype of _DV1_".

## Reasoning:

_The proposed change is the *minimal* needed change, but it would be conceptually cleaner to make the change in the rule that allows extension type to implement a **non-extension-type** supertype of its representation type, because the change is really about removing the "non-extension type" limit from **that** rule._

I expect to see extension type with `implements` clauses mainly using one of two fundamental patterns:
* Additive: Implement/expose the representation type, add more members.
* Parallel: Implement another extension type, whose representation type is a supertype of your representation type.

The pattern for adding more members to a type using extension types is:
```dart
extension type MoreFoo(Foo _) implements Foo {
  moreMembers() {}
}
```
That currently only works if `Foo` is a non-extension type. (See https://github.com/dart-lang/sdk/issues/53739).

The parallel case would be something like:
```dart
extension type FancyWidget(Widget _) {
  // something
}
extension type FancyListWidget(ListWidget _) implements FancyWidget {
 // something
}
```
where the extension types *wrap* and mirror the type hierarchy of the representation types.

I expect users to understand extension types mainly in terms of those two patterns of usage,
and I believe the two patterns should be supported equally and independently, 
meaning that an extension type `E` with representation type `R` may implement a type `T` iff:
* `T` is not one of the types we disallow explicitly (structural types, `Null`, etc.), and 
* `R` \<: `T` (Additive case) or
* `T` is an extension type with representation type `S` and `R` \<: `S` (Parallel case).

The only difference between that and what is specified today is, that today the "`R` \<: `T`" case only applies if `T` is a non-extension type. I want that to remove that restriction, by also allowing `R` \<: `T` when `T` is an extension type.

That is: The current rules restrict implementing _extension types_ to the Parallel use. You cannot additively add something to an extension type.

I think that's unnecessarily restrictive, and likely confusing to users, who will not understand why they can't use the additive pattern on an extension type, because they see no reason for that restriction (and IMO rightly so, 
because there is no technical reason, and no need to prevent the pattern of adding members to an extension type).
There is no soundness issue with allowing it, it's as sound as it is for a non-extension type representation type.

The existing rule, quoted above, _is_ useful for the parallel case. It's protective, in that it only allows looking at *one* step of the representation type chain, not the transitive representation type.
It ensures soundness, while still allowing you to implement (extend in parallel with the representation type relation) an extension type that you know the immediate representation type of.

The additive pattern doesn't look at representation type at all, and could/should therefore work for *any* type.
And as currently written, it doesn't work.

For the example above to compile, `Foo` (taking the role of both `R` and _DV1_ in the rule) must be a subtype of its own representation type, `S`, which it is _possible_ if `S` is not an extension type, and `Foo` implements precisely `S`. (It can implement any supertype of its non-extension-type representation type, so to also be a subtype, it must implement that type precisely.)
But there is no need for that.

As a concrete issue, I envision, maybe, possibly, eventually making `MapEntry` an extension type on a record.
It's currently a `final` class, so at some point, we should feel confident that there are no subtypes anywhere.
However, with the current rules, making that change would break an extension type like:
```dart
extension type FancyMapEntry<K, V>(MapEntry<K, V> _) implements MapEntry<K, V> {
  // more methods
}
```
And for no good reason, it should just keep working, because it doesn't *care* whether `MapEntry` is an extension type or not. It's just a type.

**Comment by eernstg** [comment]
Makes sense! (after some heated debates ;-) However, I think the new rule should be

> ...
> error occurs if `R` is not a subtype of `S` and `R` is not a subtype of `V1`.

**Comment by lrhn** [comment]
> new rule should be

Yes, that!

**Comment by eernstg** [comment]
I'd recommend that we accept this proposal: https://github.com/dart-lang/language/pull/3405.

**Comment by eernstg** [comment]
Closing: The proposal was accepted and #3405 was landed.
